### PR TITLE
Migrate first usable base line

### PR DESCRIPTION
**Pull request by Jooshua**
Pull Request for 

- https://github.com/Jooshua/e4CompareFramework-team2/issues/2 JavaFX Project Setup
- https://github.com/Jooshua/e4CompareFramework-team2/issues/9 Migrate ProjectExplorerKeyListener
- https://github.com/Jooshua/e4CompareFramework-team2/issues/12 Migrate SelectionListener
- https://github.com/Jooshua/e4CompareFramework-team2/issues/16 Migrate SWT Context Menu
- https://github.com/Jooshua/e4CompareFramework-team2/issues/17 Add refresh functionality to project tree

This migrates most of the ProjectExplorer to JavaFX.

### CODE REVIEW

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import javafx.embed.swt.FXCanvas;
+import javafx.embed.swt.SWTFXUtils;
+import javafx.scene.Node;
+import javafx.scene.Scene;
+import javafx.scene.control.TreeItem;
+import javafx.scene.image.ImageView;
+import javafx.scene.image.WritableImage;
+
+public class ProjectExplorerViewController {
+	private static final String PROJECT_EXPLORER_VIEW_FXML = "/ui/view/ProjectExplorerView.fxml";
```
Comment (paulo-haas) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java]: [Paths.get(...)](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Paths.html) concatenates subpaths obeying the system file separator (e.g. `/` for Linux or `\` for Windows).

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	 * @param isRoot true if a given node is a root node
+	 * @param state maps the expanded status for each node in the tree
+	 * @return a new tree item
+	 */
+	private TreeItem<FileTreeElement> buildTree(FileTreeElement parentNode, boolean isRoot, HashMap<String, Boolean> state) {
+
+		Node imgNode = isRoot ? null : getImage(parentNode);
+
+		TreeItem<FileTreeElement> currentNode = new TreeItem<FileTreeElement>(parentNode, imgNode);
+		if (state != null) {
```
Comment (paulo-haas) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java]: The if statements can be combined (`state != null && state.containsKey(...)`.  If state is null then the second condition won't be checked.

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+				// load extended file icons
+				if (fileExtensions.containsKey(fileExtension)) {
+					image = fileExtensions.get(fileExtension).getIcon(services.imageService);
+				} else if (fileExtension.equals(E4CStringTable.FILE_ENDING_XML)) {
+					image = services.imageService.getImage(null, "icons/Explorer_View/items/xml16.png");
+				} else {
+					// default file icon
+					image = services.imageService.getImage(null, "icons/Explorer_View/items/file16.png");
+				}
+				;
```
Comment (paulo-haas) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java]: Addtional semicolon.

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/ui/view/ProjectExplorerView.fxml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<?import javafx.scene.control.ContextMenu?>
+<?import javafx.scene.control.MenuItem?>
+<?import javafx.scene.control.TreeView?>
+<?import javafx.scene.layout.BorderPane?>
+
+<BorderPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="400.0" prefWidth="600.0" xmlns="http://javafx.com/javafx/8.0.171" xmlns:fx="http://javafx.com/fxml/1" fx:controller="de.tu_bs.cs.isf.e4cf.parts.project_explorer.view.ProjectExplorerView">
```
Comment (paulo-haas) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/ui/view/ProjectExplorerView.fxml:8]: Wird wohl korrekt sein, aber was bedeutet hier `-Infinity`? Ich habe leider in der Doku nichts gefunden. Hätten da nicht die default values ausgereicht?

Comment (paulo-haas) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/ui/view/ProjectExplorerView.fxml:8]: Vielleicht könnt ihr mir ja ein Link oder so schicken :sweat_smile:

**Review by paulo-haas — APPROVED** [review]

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/ui/view/ProjectExplorerView.fxml (hunk 4 above)
Comment (KamilRosiak) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/ui/view/ProjectExplorerView.fxml:8]: Das ist damit das mit der größe des Parents wächst. Das ist quasi Integer.MAX_VALUE

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+		
+		_menuService.registerContextMenu(canvas, StringTable.PROJECT_EXPLORER_CONTEXT_MENU_ID);
+
+		Scene scene = new Scene(loader.getNode());
+		canvas.setScene(scene);
+	
+		setupSelectionService();
+		
+		// listeners
+		scene.setOnKeyPressed(new ProjectExplorerKeyListener(context));
```
Comment (KamilRosiak) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java:95]: // adding key listeners to the scene

**Comment by KamilRosiak** [comment]
- Refactor the file path strings into a FileTable/StringTable 
- complete the java doc comments for methods and classes

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java (hunk 1 above)
Comment (Jooshua) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java]: Sure

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java (hunk 2 above)
Comment (Jooshua) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java]: Sure

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java (hunk 3 above)
Comment (Jooshua) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java]: Removed

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java (hunk 5 above)
Comment (Jooshua) [reply de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/ProjectExplorerViewController.java:95]: There also will be other listeners added here

**Comment by Jooshua** [comment]
> 
> 
>     * Refactor the file path strings into a FileTable/StringTable
> 
>     * complete the java doc comments for methods and classes

Resolved

**Review by paulo-haas — APPROVED** [review]

File: de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/ProjectExplorerKeyListener.java [hunk 6]
```diff
@@ -1,53 +1,81 @@
...
+	// Eclipse context for commands
+	IEclipseContext eclipseContext;
+	
+	@Inject
+	ServiceContainer services;
+	
+	// Define key combinations here
+	KeyCombination kcRename = new KeyCodeCombination(KeyCode.R, KeyCombination.CONTROL_DOWN);	
+	
+	public ProjectExplorerKeyListener(IEclipseContext  eclipseContext) {
```
Comment (Blum-P) [inline de.tu_bs.cs.isf.e4cf.parts.project_explorer/src/de/tu_bs/cs/isf/e4cf/parts/project_explorer/listeners/ProjectExplorerKeyListener.java:33]: Just a little thing: double space

**Review by Blum-P — COMMENTED** [review]
Good job. Really good code.

**Pull request merged**